Provis.UnitTests/Core/Services/TaskServiceTests.cs

```diff
@@ -0,0 +1,1234 @@
+                .Where(x => x.StatusCode == HttpStatusCode.NotFound)
+                .WithMessage(ErrorMessages.TaskNotFound);
+        }
+
+        [Test]
+        [TestCase("1")]
+        public async Task JoinTask_NoPermission_ThrowHttpException(string userId)
+        {
+            var taskMock = GetWorkspaceTask;
+            var taskAssign = AssignOnTaskDTO;
+
+            SetupTaskGetByKeyAsync(taskMock);
+            Func<Task> act = () => _taskService.JoinTaskAsync(taskAssign, userId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.Forbidden)
+                .WithMessage(ErrorMessages.NotPermissionAssignUser);
+        }
+
+        [Test]
+        [TestCase("2", false)]
+        public async Task JoinTask_UserNotMember_ThrowHttpException(string userId, bool isMember)
+        {
+            var taskMock = GetWorkspaceTask;
+            var taskAssign = AssignOnTaskDTO;
+
+            SetupTaskGetByKeyAsync(taskMock);
+            SetupUserWorkspaceAnyBySpecAsync(isMember);
+            Func<Task> act = () => _taskService.JoinTaskAsync(taskAssign, userId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.BadRequest)
+                .WithMessage(ErrorMessages.UserNotMember);
+        }
+
+        [Test]
+        [TestCase("2", true, true)]
+        public async Task JoinTask_AlreadyAssigned_ThrowHttpException(string userId, bool isMember, bool assigned)
+        {
+            var taskMock = GetWorkspaceTask;
+            var taskAssign = AssignOnTaskDTO;
+
+            SetupJoinTask(taskMock, isMember, assigned);
+
+            Func<Task> act = () => _taskService.JoinTaskAsync(taskAssign, userId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.BadRequest)
+                .WithMessage(ErrorMessages.UserAlreadyAssigned);
+        }
+
+        [Test]
+        [TestCase("2", true, false)]
+        public async Task JoinTask_AssignUser_ReturnTaskComplete(string userId, bool isMember, bool assigned)
+        {
+            var taskMock = GetWorkspaceTask;
+            var taskAssignDTO = AssignOnTaskDTO;
+            var taskAssign = new UserTask()
+            {
+                TaskId = taskAssignDTO.Id,
+                IsUserDeleted = false,
+                Task = taskMock,
+                UserId = taskAssignDTO.AssignedUser.UserId,
+                UserRoleTagId = taskAssignDTO.AssignedUser.RoleTagId
+            };
+
+            SetupJoinTask(taskMock, isMember, assigned);
+            SetupUserTaskSaveChangesAsync();
+            SetupUserTaskAddAsync(taskAssign);
+            SetupMetricsIncrement();
+            _mapperMock.SetupMap(taskAssignDTO, taskAssign);
+
+            var result = _taskService.JoinTaskAsync(taskAssignDTO, userId);
```

Comment:
Why you don't use await for async invocation

Reply:
If I use await, there wouldn't be value for the assert block